### PR TITLE
Add arithmetic to SpatialDimension

### PR DESCRIPTION
**Pull request by fzimmermann89**
Follow up to #416 with correct type hints

### CODE REVIEW

**Comment by fzimmermann89** [comment]
I also moved Rotation out of utils (as it depended on SpatialDimension) and TYPE_CHECKING guarded modify_acq_info to avoid circular imports. Functions in utils should not depend on data classes (otherwise, we can't use them in data classes)
As modify_acq_info will be deprecated anyways once we move that functionality into acqinfo, I only added the future annotations and guard.

These changes might be moved to a separate (stacked) PR .


I removed xyz to force us to always be zyx or be explicit about it and write (spatialdim.x,spatialdim.y,spatialdim.z). Otherwise, the temptation to wrongly use xyz is high..

**Comment by fzimmermann89** [comment]
This passes all tests from #416  and passes mypy

**Comment by fzimmermann89** [comment]
(To be merged after #418  into main)

File: src/mrpro/data/SpatialDimension.py [hunk 1]
```diff
@@ -101,3 +102,136 @@ def zyx(self) -> tuple[T, T, T]:
...
+    @overload
+    def __mul__(self: SpatialDimension[int], other: int | SpatialDimension[int]) -> SpatialDimension[int]: ...
+    @overload
+    def __mul__(self: SpatialDimension[int], other: float | SpatialDimension[float]) -> SpatialDimension[float]: ...
+    @overload
+    def __mul__(self: SpatialDimension[float], other: float | SpatialDimension[float]) -> SpatialDimension[float]: ...
+
+    @overload
+    def __mul__(
+        self: SpatialDimension[torch.Tensor], other: float | T | SpatialDimension
```
Comment (schuenke) [inline src/mrpro/data/SpatialDimension.py]: why do we need the an explicit `float` here? Shouldn't `float` be covered by `T` already?

File: src/mrpro/data/SpatialDimension.py [hunk 2]
```diff
@@ -101,3 +102,136 @@ def zyx(self) -> tuple[T, T, T]:
...
+    ) -> SpatialDimension[torch.Tensor]: ...
+
+    def __mul__(self: SpatialDimension[T], other: T | float | SpatialDimension) -> SpatialDimension:
```
Comment (schuenke) [inline src/mrpro/data/SpatialDimension.py]: see `float` / `T` comment above. If we do need it, using the same order everywhere would be nice. 

Please also check the rest of the code... appeards several times

**Review by schuenke — APPROVED** [review]
lgtm. IMO all cases are covered. Didn't go through all the tests, but they seem to cover everhing as well.

Please check my specific comments about the `float | T` type hints.

File: src/mrpro/data/SpatialDimension.py (hunk 2 above)
Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py]: no, it is required, you can write it out as, for example:
SpatialDimension[int], other: int | float | SpatialDimension
All spatial dimensions can be multiplied by a scalar (float, which then includes int).
But only SpatialDimension[np.ndarray] can be multiplied by np.ndarray , only SpatialDimensin[tensor] by tensor...
 
 or am I missing something?

Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py]: i can go through and sort the order

Comment (schuenke) [reply src/mrpro/data/SpatialDimension.py]: but T is `int | float | torch.Tensor` already, no?
so `float | T` is `float | int | float | torch.Tensor` , no?

Or am I actually missing something here?

Comment (schuenke) [reply src/mrpro/data/SpatialDimension.py]: ah, got it... it's needed. Sorry for the confusion! Hmm... but maybe not in the last overload... I will also check it again 😁

Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py]: but maybe your are right an the cases can be cleaned up, let me check

**Comment by fzimmermann89** [comment]
i cleaned the type hints a bit.

also, I removed the conversion option from the class methods.
Instead, there is now an apply_ that is used after creation to do the conversion.
This will anyways be required for rearranging etc..

**Comment by fzimmermann89** [comment]
Also, I added typing asserts as tests.
This is in the same test_spatialdimension.py file, even though the tests is not actually run by pytest -- just parsed by mypy.

maybe @schuenke can have another look ...

File: src/mrpro/data/SpatialDimension.py [hunk 3]
```diff
@@ -70,34 +69,342 @@ def from_array_xyz(
...
+        self: SpatialDimension[T_co_scalar] | SpatialDimension[T_co_vector],
+        other: SpatialDimension[T_co_scalar] | SpatialDimension[T_co_vector],
+    ) -> bool | T_co_vector:
+        """Check if self is greater than or equal to other."""
+        if not isinstance(other, SpatialDimension):
+            return NotImplemented
+        return (self.x >= other.x) & (self.y >= other.y) & (self.z >= other.z)
+
+    def __post_init__(self):
+        """Ensure that the data is of the same type."""
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py]: this actually checks the shape and not the type

**Comment by ckolbPTB** [comment]
Why do we want to support `np.ndarray` for SpatialDimension? It would be nice if that would be supported in the init and maybe from_-functions but it could then be converted to tensors.

**Comment by fzimmermann89** [comment]
> Why do we want to support `np.ndarray` for SpatialDimension? It would be nice if that would be supported in the init and maybe from_-functions but it could then be converted to tensors.

from_array* will convert to tensors.
I mostly added nd.array type hints to see if the hints are correct or if errors are just hidden by the fact that there is only one vector type supported.  The biggest uses I see is for compatibility and object arrays (example: array of callables).
I would leave it in as long as it doesn't cause issues.

**Comment by ckolbPTB** [comment]
> I would leave it in as long as it doesn't cause issues.

It seems strange to me why we would support ndarrays here when everywhere else we ensure torch.tensors. I would go the other way and remove it unless we really need it. Otherwise we should also support it for getitem/setitem

**Comment by fzimmermann89** [comment]
> Otherwise we should also support it for getitem/setitem

I added the overloads for getitem and setitem and a test for these.

**Comment by fzimmermann89** [comment]
> > Otherwise we should also support it for getitem/setitem
> 
> I added the overloads for getitem and setitem and a test for these.

Scratch that (_Was kümmert mich mein Geschwätz von gestern_). 
I removed np.ndarray support, as I wont use a SpatialDimension of callables in the use case I had in mind anyway.

File: src/mrpro/data/SpatialDimension.py [hunk 4]
```diff
@@ -3,16 +3,38 @@
...
 
 import numpy as np
 import torch
 from numpy.typing import ArrayLike
 
+import mrpro.utils.typing as type_utils
 from mrpro.data.MoveDataMixin import MoveDataMixin
 
-T = TypeVar('T', int, float, torch.Tensor)
+# Change here to add more types
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py:18]: could we add a sentence why we need the types to be covariant?

File: src/mrpro/data/SpatialDimension.py [hunk 5]
```diff
@@ -3,16 +3,38 @@
...
+T = TypeVar('T', torch.Tensor, int, float)
+T_co = TypeVar('T_co', torch.Tensor, int, float, covariant=True)
+T_co_float = TypeVar('T_co_float', float, torch.Tensor, covariant=True)
+T_co_vector = torch.Tensor
+T_co_scalar = TypeVar('T_co_scalar', int, float, covariant=True)
+
+
+def _as_vectortype(x: ArrayLike) -> VectorTypes:
+    """Convert ArrayLike to VectorType."""
+    if isinstance(x, VectorTypes) and type(x) in get_args(VectorTypes):
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py:32]: is `type(x) in get_args(VectorTypes)` not enough here?

File: src/mrpro/data/SpatialDimension.py [hunk 6]
```diff
@@ -24,80 +46,401 @@ class XYZ(Protocol[T]):
...
         data
             shape (..., 3) in the order (x,y,z)
-        conversion
-            will be called for each value to convert it
         """
-        if not isinstance(data, np.ndarray | torch.Tensor):
-            data = np.asarray(data)
-
+        data_ = _as_vectortype(data)
         if np.size(data, -1) != 3:
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py]: I would have this check before _as_vectortype

File: src/mrpro/data/SpatialDimension.py [hunk 7]
```diff
@@ -24,80 +46,401 @@ class XYZ(Protocol[T]):
...
         ----------
         data
             shape (..., 3) in the order (z,y,x)
-        conversion
-            will be called for each value to convert it
         """
-        data = torch.flip(torch.as_tensor(data), (-1,))
-        return SpatialDimension.from_array_xyz(data, conversion)
+        data = _as_vectortype(data)
+        if np.size(data, -1) != 3:
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py]: same as above

File: src/mrpro/data/SpatialDimension.py [hunk 8]
```diff
@@ -24,80 +46,401 @@ class XYZ(Protocol[T]):
...
+        Returns
+        -------
+            Empty tuple if x, y, and z are scalar types, otherwise shape
+
+        Raises
+        ------
+            ValueError if the shapes are not equal
+        """
+        if isinstance(self.x, ScalarTypes) and isinstance(self.y, ScalarTypes) and isinstance(self.z, ScalarTypes):
+            return ()
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py:440]: would I not expect an error if I call shape on a scalar?

**Review by ckolbPTB — APPROVED** [review]
some minor last comments

File: src/mrpro/data/SpatialDimension.py (hunk 5 above)
Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py:32]: At runtime yes. But mypy curently only understands the former.

File: src/mrpro/data/SpatialDimension.py (hunk 6 above)
Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py]: It should be this order.

ArrayLike is anything that numpy can convert to an Array, for example a list of lists.
Np.shape on objects that do not have .shape is plemented by converting to an Array, then checking it's size.
So switching the commands here would result in double conversion and has no advantages?

File: src/mrpro/data/SpatialDimension.py (hunk 8 above)
Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py:440]: Calling shape on scalar numpy arrays or tensors returns also an empty tuple.

File: src/mrpro/data/SpatialDimension.py (hunk 6 above)
Comment (ckolbPTB) [reply src/mrpro/data/SpatialDimension.py]: should it then be np.size(data_, -1)?

Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py]: YES! 
thank you

**Comment by fzimmermann89** [comment]
pre-commit.ci autofix

**Comment by fzimmermann89** [comment]
pre-commit.ci autofix

**Pull request merged**